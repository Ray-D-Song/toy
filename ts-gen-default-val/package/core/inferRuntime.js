export function inferDefaultValue(node, declaredTypes) {
  switch (node.type) {
    case 'TSStringKeyword':
      return `''`
    case 'TSNumberKeyword':
      return '0'
    case 'TSBooleanKeyword':
      return 'false'
    case 'TSObjectKeyword':
      return '{}'
    case 'TSTypeLiteral':
      // TODO recursive calls
      return '{}'
    case 'TSFunctionType':
      return 'function() {}'
    case 'TSArrayType':
      return '[]'
    case 'TSTupleType':
      return '[]'

    case 'TSLiteralType':
      switch (node.literal.type) {
        case 'StringLiteral':
          return ['String']
        case 'BooleanLiteral':
          return ['Boolean']
        case 'NumericLiteral':
        case 'BigIntLiteral':
          return ['Number']
        default:
          return [`null`]
      }

    case 'TSTypeReference':
      if (node.typeName.type === 'Identifier') {
        if (declaredTypes[node.typeName.name]) {
          return declaredTypes[node.typeName.name]
        }
        switch (node.typeName.name) {
          case 'Array':
          case 'Function':
          case 'Object':
          case 'Set':
          case 'Map':
          case 'WeakSet':
          case 'WeakMap':
            return [node.typeName.name]
          case 'Record':
          case 'Partial':
          case 'Readonly':
          case 'Pick':
          case 'Omit':
          case 'Exclude':
          case 'Extract':
          case 'Required':
          case 'InstanceType':
            return ['Object']
        }
      }
      return [`null`]

    case 'TSUnionType':
      return [
        ...new Set(
          [].concat(node.types.map(t =>
            inferRuntimeType(t, declaredTypes)
          ))
        )
      ]

    case 'TSIntersectionType':
      return ['Object']

    default:
      return [`null`] // no runtime check
  }
}