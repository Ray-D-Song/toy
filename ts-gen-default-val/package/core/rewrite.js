export function rewriteSourceCode(source, reg, list) {
  let idx = 0
  const res = source.replace(reg, (m) => {
    let replacement = list[idx]
    idx = (idx + 1) % list.length
    return replacement
  })
  return res
}