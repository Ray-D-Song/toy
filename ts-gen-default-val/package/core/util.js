import fs from 'fs'

const logFilePath = '/Users/ray-d-song/codebase/toy/ts-gen-default-val/dist/demo.log'

Date.prototype.format = function(fmt) { 
  var o = { 
     "M+" : this.getMonth()+1,                 //月份 
     "d+" : this.getDate(),                    //日 
     "h+" : this.getHours(),                   //小时 
     "m+" : this.getMinutes(),                 //分 
     "s+" : this.getSeconds(),                 //秒 
     "q+" : Math.floor((this.getMonth()+3)/3), //季度 
     "S"  : this.getMilliseconds()             //毫秒 
 }; 
 if(/(y+)/.test(fmt)) {
         fmt=fmt.replace(RegExp.$1, (this.getFullYear()+"").substr(4 - RegExp.$1.length)); 
 }
  for(var k in o) {
     if(new RegExp("("+ k +")").test(fmt)){
          fmt = fmt.replace(RegExp.$1, (RegExp.$1.length==1) ? (o[k]) : (("00"+ o[k]).substr((""+ o[k]).length)));
      }
  }
 return fmt; 
}    

export function recordLog(content) {
  const logContent = `
${new Date().format('yyyy-MM-dd hh:mm:ss')}
${JSON.stringify(content, '', '\t')}

  `
  fs.appendFileSync(logFilePath, logContent)
}

export function logWarn(arg) {
  console.log('\x1b[33m%s\x1b[0m', arg)
  recordLog(arg)
}

export function logError(arg) {
  console.log('\x1b[31m%s\x1b[0m', arg)
  recordLog(arg)
}