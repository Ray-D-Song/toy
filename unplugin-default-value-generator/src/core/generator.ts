import { parse as babelParse } from '@babel/parser'
import type { Statement, TSType } from '@babel/types'
import type { Options } from '../types'

function extractCodeBlock(reg: RegExp, raw: string) {
	return raw.match(reg)
}

export function extractTypeParameters(meta: Statement) {
	if (meta.type !== 'ExpressionStatement')
		throw new Error('macro should be called with expressions')
	if (meta.expression.type !== 'CallExpression')
		throw new Error('macro should be called with expressions')
	if (!meta.expression.typeParameters)
		throw new Error('macro calls must input type parameter')
	if (meta.expression.typeParameters.params.length !== 1)
		throw new Error('macro can only have 1 type parameter')

	return meta.expression.typeParameters.params[0]
}

export function inferDefaultValue(typeParamAST: TSType) {
	switch (typeParamAST.type) {
		case 'TSStringKeyword':
			return `''`
		case 'TSNumberKeyword':
			return '0'
		case 'TSBooleanKeyword':
			return 'false'
    case 'TSObjectKeyword':
      return '{}'
    case 'TSNullKeyword':
      return 'null'
    case 'TSTypeLiteral':
      return handleLiteralType:
	}
}

export function parse(code: string) {
	const meta = babelParse(code, {
		plugins: ['typescript'],
	}).program.body[0]

	const typeParam = extractTypeParameters(meta)

	return typeParam
}

export default function createGenerator(options: Options) {
	const macroNameReg = new RegExp(`${options.macroName}<([\s\S]*)+>\(\)`, 'g')

	function generateDefaultValue(raw: string) {
		const codeBlockList = extractCodeBlock(macroNameReg, raw)
		if (!codeBlockList)
			return raw
		for (const code of codeBlockList)
			parse(code)
	}

	return generateDefaultValue
}
