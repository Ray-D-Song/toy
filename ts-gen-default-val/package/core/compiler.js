import pkg from '@babel/parser'
import { logWarn, logError } from './util.js'
import { inferDefaultValue } from './inferRuntime.js'
import { rewriteSourceCode } from './rewrite.js'

const { parse: babelParse, ParserOptions, ParserPlugin } = pkg

const reg = /createDefault<([\s\S]*)+>\(\)/g;

function extractParameters(paramMembers) {
  let valueStr = '{'
  for(const p of paramMembers) {
    const keyName = p.key.name
    const defaultValue = inferDefaultValue(p.typeAnnotation.typeAnnotation)
    const str = `${keyName}: ${defaultValue},`
    valueStr = valueStr + str
  }
  valueStr = valueStr + '}'
  return valueStr
}

export default function compile(str, id) {
  if(!reg.test(str)) {
    logWarn('method not detected')
    return str
  }

  let generateCodeBlocks = []
  const matches = str.match(reg)
  try {
    for(const m of matches) {
      const meta = babelParse(m, {
        plugins: ['typescript']
      }).program.body[0]

      if(!meta.expression.typeParameters) 
        throw logError('type parameters not exist')
      if(meta.expression.typeParameters.params.length > 1)
        throw logError('accept only one parameter.')
      const param = meta.expression.typeParameters.params[0]
      if(param.type !== 'TSTypeLiteral')
        throw logError('type can only be TSTypeLiteral')
      const paramMembers = param.members
      const extractResults = extractParameters(paramMembers)
      generateCodeBlocks.push(extractResults)
    }

    return rewriteSourceCode(str, reg, generateCodeBlocks)

  } catch (e) {
    logError('compile error: \n')
    logError(e)
  }
}