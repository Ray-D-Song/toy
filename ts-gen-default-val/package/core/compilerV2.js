import pkg from '@babel/parser'
import { logWarn, logError } from './util.js'
import { inferDefaultValue } from './inferRuntime.js'
import { rewriteSourceCode } from './rewrite.js'

const { parse: babelParse, ParserOptions, ParserPlugin } = pkg

function extractParameters(paramMembers) {
  let valueStr = '{'
  for(const p of paramMembers) {
    const keyName = p.key.name
    const defaultValue = inferDefaultValue(p.typeAnnotation.typeAnnotation)
    const str = `${keyName}: ${defaultValue},`
    valueStr = valueStr + str
  }
  valueStr = valueStr + '}'
  return valueStr
}

export default function compile(str, id) {
  try {
    // const meta = babelParse(str, {
    //   plugins: ['typescript']
    // })

    
    return str

  } catch (e) {
    logError('compile error: \n')
    logError(e)
  }
}