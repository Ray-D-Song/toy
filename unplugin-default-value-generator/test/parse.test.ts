import { expect, it } from 'vitest'
import { parse } from '../src/core/generator'
import { notypeArgInput_1, notypeArgInput_2, notypeArgInput_3, parseInput, singleTypeParseInput } from './fixtures/parse'

const path = './__snapshots__/parse.test.ts.snap'

it('parse code return meta data', () => {
	expect(parse(parseInput)).toMatchSnapshot(path)
	expect(parse(singleTypeParseInput)).toMatchSnapshot(path)
})

it('no type arg input, should throw err', () => {
	expect(() => parse(notypeArgInput_1)).toThrowError('macro should be called with expressions')
	expect(() => parse(notypeArgInput_2)).toThrowError('macro calls must input type parameter')
	expect(() => parse(notypeArgInput_3)).toThrowError()
})
