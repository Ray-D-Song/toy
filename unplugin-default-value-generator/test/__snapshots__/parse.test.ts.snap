// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse code return meta data > ./__snapshots__/parse.test.ts.snap 1`] = `
Node {
  "end": 47,
  "loc": SourceLocation {
    "end": Position {
      "column": 1,
      "index": 47,
      "line": 5,
    },
    "filename": undefined,
    "identifierName": undefined,
    "start": Position {
      "column": 14,
      "index": 15,
      "line": 2,
    },
  },
  "members": [
    Node {
      "computed": false,
      "end": 31,
      "key": Node {
        "end": 23,
        "loc": SourceLocation {
          "end": Position {
            "column": 6,
            "index": 23,
            "line": 3,
          },
          "filename": undefined,
          "identifierName": "name",
          "start": Position {
            "column": 2,
            "index": 19,
            "line": 3,
          },
        },
        "name": "name",
        "start": 19,
        "type": "Identifier",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 14,
          "index": 31,
          "line": 3,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 2,
          "index": 19,
          "line": 3,
        },
      },
      "start": 19,
      "type": "TSPropertySignature",
      "typeAnnotation": Node {
        "end": 31,
        "loc": SourceLocation {
          "end": Position {
            "column": 14,
            "index": 31,
            "line": 3,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 6,
            "index": 23,
            "line": 3,
          },
        },
        "start": 23,
        "type": "TSTypeAnnotation",
        "typeAnnotation": Node {
          "end": 31,
          "loc": SourceLocation {
            "end": Position {
              "column": 14,
              "index": 31,
              "line": 3,
            },
            "filename": undefined,
            "identifierName": undefined,
            "start": Position {
              "column": 8,
              "index": 25,
              "line": 3,
            },
          },
          "start": 25,
          "type": "TSStringKeyword",
        },
      },
    },
    Node {
      "computed": false,
      "end": 45,
      "key": Node {
        "end": 37,
        "loc": SourceLocation {
          "end": Position {
            "column": 5,
            "index": 37,
            "line": 4,
          },
          "filename": undefined,
          "identifierName": "age",
          "start": Position {
            "column": 2,
            "index": 34,
            "line": 4,
          },
        },
        "name": "age",
        "start": 34,
        "type": "Identifier",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 13,
          "index": 45,
          "line": 4,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 2,
          "index": 34,
          "line": 4,
        },
      },
      "start": 34,
      "type": "TSPropertySignature",
      "typeAnnotation": Node {
        "end": 45,
        "loc": SourceLocation {
          "end": Position {
            "column": 13,
            "index": 45,
            "line": 4,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 5,
            "index": 37,
            "line": 4,
          },
        },
        "start": 37,
        "type": "TSTypeAnnotation",
        "typeAnnotation": Node {
          "end": 45,
          "loc": SourceLocation {
            "end": Position {
              "column": 13,
              "index": 45,
              "line": 4,
            },
            "filename": undefined,
            "identifierName": undefined,
            "start": Position {
              "column": 7,
              "index": 39,
              "line": 4,
            },
          },
          "start": 39,
          "type": "TSNumberKeyword",
        },
      },
    },
  ],
  "start": 15,
  "type": "TSTypeLiteral",
}
`;

exports[`parse code return meta data > ./__snapshots__/parse.test.ts.snap 2`] = `
Node {
  "end": 21,
  "loc": SourceLocation {
    "end": Position {
      "column": 20,
      "index": 21,
      "line": 2,
    },
    "filename": undefined,
    "identifierName": undefined,
    "start": Position {
      "column": 14,
      "index": 15,
      "line": 2,
    },
  },
  "start": 15,
  "type": "TSStringKeyword",
}
`;
